### ISSUE TITLE
[Feature] Stratified sandboxes

### ISSUE THREAD

**Issue by dginev** [dginev/CorTeX#46]
A very natural need when working on a large corpus for iterative development is to experiment with convenient subsets of the large collection, based on certain metrics of interest.

I have become quite fond of the term "stratify", as employed by ML tools such as sklearn, so I would like to start calling these "stratified sandboxes".

In essence, cortex should provide a mechanism on each (corpus,service) report page to request a "stratified sandbox" from a finished service run. That mechanism is likely just a simple html form, allowing to select for various types of stratification:
  - by path fragments: try to provide an equal number of entries from each corpus segment (be it subdirectory, or filename globs, etc)
 - by messages: try to provide an equal number of entries for each message `severity:category:what` class
 - by count/frequency: both lower/upper bounds for the main stratification criteria. For example:
    - only 1 entry for each message registered
    - not more than 1,000 per class, stratified by path fragments
    - unlimited, for a specific message class
    - and so on

The implementation is somewhat direct. A new corpus is instantiated, named after the corpus, service, stratification nickname, and timestamp. A metadata structure is then constructed via a specialized query for postgres. Each reported entry is then added to the new corpus, and finally the service in question is registered on the new sandbox, preparing it for a conversion run.

Add to this feature the ability to download custom corpora (by receiving a dangling link and having a background task creating the download for you, at which point the link activates and optionally you get a notification), and you get a full lifecycle UX for sandboxes.

**Comment by dginev** [comment]
### Status: partially addressed — plan-of-record landed, implementation deferred

A re-evaluation against this issue's full scope. The **carve primitive** shipped 2026-06-15 (`src/backend/sandbox.rs`, `POST /api/corpora/<parent>/sandbox`, the `corpus_sandbox` job): a real corpus + `parent_corpus_id` + `selection` JSONB, a one-transaction server-side `INSERT … SELECT`, and output isolation. But that `selection` is a flat **condition filter** (`status` ∩ message `severity/category/what` ∩ `entry LIKE` ∩ a single global cap) — it is **not** a stratified sample. So the defining capability of this issue is still open.

Scorecard vs. the issue:

| Scope | State |
|---|---|
| A. Form on the **(corpus, service) report page** | ❌ form is on the corpus page only |
| B1. Stratify by **path fragments** | ❌ only `entry LIKE` |
| B2. Stratify by **message class** (`severity:category:what`) | ❌ keeps *all* matches |
| B3. **Lower/upper count bounds per class** | 🟡 only one global cap |
| C1. Auto-name `{corpus·service·nickname·timestamp}` | 🟡 user types the name |
| C2. Specialized stratified SQL | ❌ plain `EXISTS` + `LIKE` + `LIMIT` |
| C3. Each reported entry added | ✅ |
| C4. Service auto-activated to prime a run | ❌ separate manual call |
| D. Downloadable custom corpora (dangling link + bg task) | ❌ `export-dataset` is HTML results, not the corpus |

**Key design insight:** the filter and the stratification *axis* are independent — the existing fields pick the candidate pool, then one new `stratify_by` axis + per-class bounds balances it. That covers all three examples in the issue ("1 per message" = MessageClass, max 1; "≤1000 by path fragment" = PathSegment, max 1000; "unlimited for a class" = no cap). It needs **no schema migration** — every new field rides inside the existing `selection JSONB`.

Full design + owner decisions are now the plan-of-record in [`docs/STRATIFIED_SANDBOXES.md`](https://github.com/dginev/CorTeX/blob/master/docs/STRATIFIED_SANDBOXES.md). Decisions locked: floor-guarantee lower bound (two-pass under a global cap), seeded-shuffle sampling by default (reproducible), Nth `/`-segment path key. Phasing: **(1)** model + stratified windowed SQL, **(2)** auto-naming + auto-activation, **(3)** report-page form, **(4)** download lifecycle. Substantially closeable after phase 3.

Keeping this open as the tracking issue; implementation to follow.
